### PR TITLE
fix: prevent error during transformation when object has property with Promise

### PR DESCRIPTION
**Pull request by Hossein-s**
Class transformer throws error when class type has promise field.

fixes: #261

### CODE REVIEW

**Comment by chrisdevereux** [comment]
Would be great to have this merged

**Comment by Hossein-s** [comment]
@chrisdevereux 

Seems this project is not being maintained for now. If you really need this (like me :D) you can `npm https://github.com/Hossein-s/class-transformer-fork-build.git` to install from my build or you can build my fork by yourself.

**Comment by chrisdevereux** [comment]
This bit of evilness did it for me:

````
// HACK: Workaround until: https://github.com/typestack/class-transformer/pull/262
invariant(
  require('class-transformer/package.json').version === '0.2.3'
  'Patched method assumes exact dependency version'
)
patchMethod(
  require('class-transformer/TransformOperationExecutor').TransformOperationExecutor,
  'transform',
  transform =>
    function(this: unknown, parent: unknown, value: unknown) {
      if (value && isPromiseAlike(value)) {
        return value
      }

      return transform.apply(this, arguments)
    },
)
````

**Comment by PachowStudios** [comment]
@chrisdevereux Brilliant. I feel so dirty putting this in our code.

BTW for everyone's reference: `invariant` comes from the package by the same name, and `patchMethod` comes from zone.js.

**Comment by vicdup** [comment]
Hi @Hossein-s,

we are currently facing a similar problem, when using class-transformer with typeorm and nestJs. 

We are struggling on how to make class-transformer await a property when this property is exposed. Currently it result in a null object.

For instance in our entity "Booking" we have a property "Contact" defined as follow:
```
@ManyToOne(type => Contact, contact => contact.bookings, { eager: true })
 @JoinColumn()
 @Expose({ groups: ['vero'] })
 @Field(type => Contact)
 contact: Promise<Contact>;
```

when we use plain to class on a booking fetched with "findOne" of Typeorm it results in :

```
contact: {}
```

would you know how to proceed in such a case to get the serialization to await contact?

I don't think this fix works as we still get a promise after the fix and the promise is not awaited before serialization. Also with your fix we loose circle dependencies

**Comment by Hossein-s** [comment]
Hi @vicdup  😃 

This is the intended behaviour. This fix doesn't resolve promises, it just solves error that class-transformer throws when you have promise field.

Just as a suggestion, if you want to lazy load relation, then you should remove eager: true, If you want to eager load relation, then you can use `Contact` type, no need for promise.

To resolve promises you should use a utility function, something like this...
<script src="https://gist.github.com/Hossein-s/93347750c185ff2ea9a9649bb198195a.js"></script>

**Comment by PachowStudios** [comment]
@vicdup Side note, the author of typeorm said that he's planning on removing lazy-loading in 0.4.0 and he doesn't recommend using it anymore as it has a lot of issues.

**Comment by PachowStudios** [comment]
Any update on this? Would be nice to have merged soon :)

**Comment by PachowStudios** [comment]
Is this repo dead? Coming up on a year for this PR.

@Hossein-s I read your earlier response, but this is still a problem for input types when using graphql-upload with NestJS.

**Comment by Hossein-s** [comment]
> Is this repo dead? Coming up on a year for this PR.
> 
> @Hossein-s I read your earlier response, but this is still a problem for input types when using graphql-upload with NestJS.

I'm using NestJS and graphql too without problem, what is the error that you get?

**Comment by PachowStudios** [comment]
@Hossein-s This is what I have in my code:
```ts
import { GraphQLUpload } from 'apollo-server-core';
import { Exclude } from 'class-transformer';
import { Length } from 'class-validator';
import { FileUpload } from 'graphql-upload';
import { Field, ID, InputType } from 'type-graphql';

@InputType()
export class UploadDocumentInput {
  @Field()
  @Length(2, 100)
  name!: string;

  @Field(() => GraphQLUpload)
  // Promises break class-transformer
  // TODO: Remove @Exclude once https://github.com/typestack/class-transformer/pull/262 is merged
  @Exclude()
  file!: Promise<FileUpload>;
}
```

Without that `@Exclude()` I get `TypeError: Promise resolver undefined is not a function`. I'm using the NestJS `ValidationPipe`.

File: package.json [hunk 1]
```diff
@@ -24,7 +24,9 @@
     "object-to-class",
     "typescript-serializer"
   ],
-  "dependencies": {},
+  "dependencies": {
```
Comment (NoNameProvided) [inline package.json:27]: Please remove this dependency and add this functionality as a local util function.

```ts
export function isPromise<T>(p: any): p is Promise<T> {
  return p !== null && typeof p === 'object' && typeof p.then === 'function';
}
```

**Comment by NoNameProvided** [comment]
Hi! 

Thanks for the PR and sorry for the delayed reply. Please rebase the PR against the current develop branch and modify the PR according to the requested change. 

We have 
  - moved to Jest so you need to adapt your tests (it's straightforward)
  - added Prettier so you need to format your code with `npm run prettier:fix`

File: package.json (hunk 1 above)
Comment (Hossein-s) [reply package.json]: Yeah, okay.

**Comment by jvvcn** [comment]
Hi, are there any updates on this PR?

**Comment by cloudratha** [comment]
Can I pay for this fix??

**Comment by ORzazade** [comment]
> This bit of evilness did it for me:
> 
> ```ts
> // HACK: Workaround until: https://github.com/typestack/class-transformer/pull/262
> invariant(
>   require('class-transformer/package.json').version === '0.2.3',
>   'Patched method assumes exact dependency version'
> )
> patchMethod(
>   require('class-transformer/TransformOperationExecutor').TransformOperationExecutor,
>   'transform',
>   transform =>
>     function(this: unknown, parent: unknown, value: unknown) {
>       if (value && isPromiseAlike(value)) {
>         return value
>       }
> 
>       return transform.apply(this, arguments)
>     },
> )
> ```

This saved my day! 

BTW: you can use `patch-method` npm

**Comment by WillSquire** [comment]
If you want to fix this using [patch-package](https://www.npmjs.com/package/patch-package) instead, add the following to `patches/class-transformer+0.3.1.patch`:

```patch
diff --git a/node_modules/class-transformer/TransformOperationExecutor.js b/node_modules/class-transformer/TransformOperationExecutor.js
index dec36c7..81c2d09 100644
--- a/node_modules/class-transformer/TransformOperationExecutor.js
+++ b/node_modules/class-transformer/TransformOperationExecutor.js
@@ -117,6 +117,9 @@ class TransformOperationExecutor {
                 return value;
             return Buffer.from(value);
         }
+        else if (Boolean(value && typeof value.then === 'function')) {
+            return value
+        }
         else if (typeof value === 'object' && value !== null) {
             // try to guess the type
             if (!targetType && value.constructor !== Object /* && TransformationType === TransformationType.CLASS_TO_PLAIN*/)
```

**Review by NoNameProvided — DISMISSED** [review]

**Comment by NoNameProvided** [comment]
Related tracking issue: #549

**Pull request merged**